src/cockpit/protocol.py

```diff
@@ -191,21 +189,10 @@ def write_channel_data(self, channel, payload):
         else:
             logger.debug('cannot write to closed transport')
 
-    def write_message(self, _channel, **kwargs):
-        """Format kwargs as a JSON blob and send as a message
-           Any kwargs with '_' in their names will be converted to '-'
-        """
-        for name in list(kwargs):
-            if '_' in name:
-                kwargs[name.replace('_', '-')] = kwargs[name]
-                del kwargs[name]
-
-        logger.debug('sending message %s %s', _channel, kwargs)
-        pretty = CockpitProtocol.json_encoder.encode(kwargs) + '\n'
-        self.write_channel_data(_channel, pretty.encode('utf-8'))
-
-    def write_control(self, **kwargs):
-        self.write_message('', **kwargs)
+    def write_control(self, _msg: 'JsonObject | None' = None, **kwargs: JsonDocument) -> None:
+        """Write a control message.  See print_object() for details."""
```

Comment:
This refers you to print_object() to read about exactly what happens (and the docstring there describe it in a fair amount of detail).

As for the use of the name `_msg`, described in the commit message, we could indeed explain in the docstring why we used this particular name and how it might be nice to change to `/` in the future... it seemed like something that's a better fit for a commit message for me, but I guess we're less likely to lose track of it if it's a comment.

Reply:
Right, comment, not docstring (although none of that is user documentation, so doesn't matter much).

Reply:
The one thing that's difficult about this is that it's a convention that's now used in 5 different functions, and it's not clear which one should get the comment.  All of them?

I feel like the commit message is the right thing here...

Reply:
Ack, let's not dwell on this.